sharded_queue/drivers/tagged.lua

```diff
@@ -0,0 +1,168 @@
+local state = require('sharded_queue.state')
+local utils = require('sharded_queue.utils')
+local log = require('log') -- luacheck: ignore
+
+local method = {}
+
+local function tube_create(opts)
+    local space_opts = {}
+    local if_not_exists = opts.if_not_exists or false
+    space_opts.temporary = opts.temporary or false
+    space_opts.if_not_exists = if_not_exists
+    space_opts.engine = opts.engine or 'memtx'
+    space_opts.format = {
+        { name = 'task_id', type = 'unsigned' },
+        { name = 'bucket_id', type = 'unsigned' },
+        { name = 'status', type = 'string' },
+        { name = 'data', type = '*' },
+        { name = 'index', type = 'unsigned' },
+        { name = 'tags', type = 'array'}
+    }
+
+    local space = box.schema.create_space(opts.name, space_opts)
+    space:create_index('task_id', {
+        type = 'tree',
+        parts = { 'task_id' },
+        if_not_exists = if_not_exists
+    })
+    space:create_index('status', {
+        type = 'tree',
+        parts = { 'status', 'task_id' },
+        if_not_exists = if_not_exists
+    })
+
+    space:create_index('bucket_id', {
+        parts = { 'bucket_id' },
+        unique = false,
+        if_not_exists = if_not_exists
+    })
+
+    space:create_index('idx', {
+        type = 'tree',
+        parts = { 'bucket_id', 'index' },
+        if_not_exists = if_not_exists
+    })
+
+    return space
+end
+
+local function tube_drop(tube_name)
+    box.space._stat:delete(tube_name)
+    box.space[tube_name]:drop()
+end
+
+local function get_space(args)
+    return box.space[args.tube_name]
+end
+
+local function get_index(args)
+    local task = get_space(args).index.idx:max { args.bucket_id }
+    if not task or task.bucket_id ~= args.bucket_id then
+        return 1
+    else
+        return task.index + 1
+    end
+end
+
+local function normalize_task(task)
+    if task == nil then return nil end
+    return { task.task_id, task.status, task.data }
+end
+
+-- put task in space
+function method.put(args)
+    local idx = get_index(args)
+    local task_id = utils.pack_task_id(args.bucket_id, args.bucket_count, idx)
+    local task = get_space(args):insert { task_id, args.bucket_id, state.READY, args.data, idx, args.tags }
+    return normalize_task(task)
+end
+
+local function tags_are_available(task_tags, take_tags)
+    if #task_tags == 0 then
+        return #take_tags == 0
+    end
+
+    local task_tags_set, take_tags_set = {}, {}
+    for _, tag in ipairs(task_tags) do
+        task_tags_set[tag] = true
+    end
+
+    for _, tag in ipairs(take_tags) do
+        take_tags_set[tag] = true
+    end
+
+    for tag, _ in pairs(task_tags_set) do
+        if take_tags_set[tag] == nil then
+            return false
+        end
+    end
+    return true
+end
+
+-- take task
+function method.take(args)
```

Comment:
A single "unsuitable" task will block all others. You need to scan space searching for suitable tasks.